File: src/main/java/in/dragonbra/javasteam/steam/handlers/steamunifiedmessages/SteamUnifiedMessages.java
package in.dragonbra.javasteam.steam.handlers.steamunifiedmessages;

import com.google.protobuf.AbstractMessage;
import com.google.protobuf.GeneratedMessage;
import in.dragonbra.javasteam.base.ClientMsgProtobuf;
import in.dragonbra.javasteam.base.IPacketMsg;
import in.dragonbra.javasteam.base.PacketClientMsgProtobuf;
import in.dragonbra.javasteam.enums.EMsg;
import in.dragonbra.javasteam.handlers.ClientMsgHandler;
import in.dragonbra.javasteam.steam.handlers.steamunifiedmessages.callback.ServiceMethodNotification;
import in.dragonbra.javasteam.steam.handlers.steamunifiedmessages.callback.ServiceMethodResponse;
import in.dragonbra.javasteam.types.AsyncJobSingle;
import in.dragonbra.javasteam.types.JobID;
import in.dragonbra.javasteam.util.Strings;
import in.dragonbra.javasteam.util.compat.Consumer;
import in.dragonbra.javasteam.util.log.LogManager;
import in.dragonbra.javasteam.util.log.Logger;

import java.lang.reflect.Method;
import java.util.Collections;
import java.util.HashMap;
import java.util.Map;

/**
 * @author Lossy
 * @since 2023-01-04
 * <p>
 * This handler is used for interacting with Steamworks unified messaging
 */
public class SteamUnifiedMessages extends ClientMsgHandler {

    private static final Logger logger = LogManager.getLogger(SteamUnifiedMessages.class);
    private Map<EMsg, Consumer<IPacketMsg>> dispatchMap;

    public SteamUnifiedMessages() {
        dispatchMap = new HashMap<>();

        dispatchMap.put(EMsg.ServiceMethodResponse, this::handleServiceMethodResponse);
        dispatchMap.put(EMsg.ServiceMethod, this::handleServiceMethod);

        dispatchMap = Collections.unmodifiableMap(dispatchMap);
    }

    @Override
    public void handleMsg(IPacketMsg packetMsg) {
        if (packetMsg == null) {
            throw new IllegalArgumentException("packetMsg is null");
        }

        Consumer<IPacketMsg> dispatcher = dispatchMap.get(packetMsg.getMsgType());
        if (dispatcher != null) {
            dispatcher.accept(packetMsg);
        }
    }

    /**
     * Sends a message.
     * Results are returned in a {@link in.dragonbra.javasteam.steam.handlers.steamunifiedmessages.callback.ServiceMethodResponse}.
     *
     * @param rpcName    Name of the RPC endpoint. Takes the format ServiceName.RpcName
     * @param message    The message to send.
     * @param <TRequest> The type of protobuf object.
     * @return The JobID of the request. This can be used to find the appropriate {@link in.dragonbra.javasteam.steam.handlers.steamunifiedmessages.callback.ServiceMethodResponse}.
     */
    public <TRequest extends GeneratedMessage.Builder<TRequest>> AsyncJobSingle<ServiceMethodResponse> sendMessage(String rpcName, GeneratedMessage message) {
        if (message == null) {
            throw new IllegalArgumentException("message is null");
        }

        JobID jobID = client.getNextJobID();
        EMsg eMsg = client.getSteamID() == null ? EMsg.ServiceMethodCallFromClientNonAuthed : EMsg.ServiceMethodCallFromClient;

        ClientMsgProtobuf<TRequest> msg = new ClientMsgProtobuf<>(message.getClass(), eMsg);
        msg.setSourceJobID(jobID);
        msg.getHeader().getProto().setTargetJobName(rpcName);
        msg.getBody().mergeFrom(message);

        client.send(msg);

        return new AsyncJobSingle<>(client, jobID);
    }

    /**
     * Sends a notification.
     *
     * @param rpcName    Name of the RPC endpoint. Takes the format ServiceName.RpcName
     * @param message    The message to send.
     * @param <TRequest> The type of protobuf object.
     */
    public <TRequest extends GeneratedMessage.Builder<TRequest>> void sendNotification(String rpcName, GeneratedMessage message) {
        if (message == null) {
            throw new IllegalArgumentException("message is null");
        }

        EMsg eMsg = client.getSteamID() == null ? EMsg.ServiceMethodCallFromClientNonAuthed : EMsg.ServiceMethodCallFromClient;
        ClientMsgProtobuf<TRequest> msg = new ClientMsgProtobuf<>(message.getClass(), eMsg);
        msg.getHeader().getProto().setTargetJobName(rpcName);
        msg.getBody().mergeFrom(message);

        client.send(msg);
    }

    private void handleServiceMethodResponse(IPacketMsg packetMsg) {
        if (!(packetMsg instanceof PacketClientMsgProtobuf)) {
            throw new IllegalArgumentException("Packet message is expected to be protobuf.");
        }

        PacketClientMsgProtobuf packetMsgProto = (PacketClientMsgProtobuf) packetMsg;

        client.postCallback(new ServiceMethodResponse(packetMsgProto));
    }

    @SuppressWarnings("unchecked")
    private void handleServiceMethod(IPacketMsg packetMsg) {
        if (!(packetMsg instanceof PacketClientMsgProtobuf)) {
            throw new IllegalArgumentException("Packet message is expected to be protobuf.");
        }

        PacketClientMsgProtobuf packetMsgProto = (PacketClientMsgProtobuf) packetMsg;

        String jobName = packetMsgProto.getHeader().getProto().getTargetJobName();
        if (!Strings.isNullOrEmpty(jobName)) {
            String[] splitByDot = jobName.split("\\.");
            String[] splitByHash = splitByDot[1].split("#");

            String serviceName = splitByDot[0];
            String methodName = splitByHash[0];

            String serviceInterfaceName = "in.dragonbra.javasteam.rpc.interfaces.I" + serviceName;
            try {
                logger.debug("Handling Service Method: " + serviceInterfaceName);

                Class<?> serviceInterfaceType = Class.forName(serviceInterfaceName);

                Method method = null;
                for (Method m : serviceInterfaceType.getDeclaredMethods()) {
                    if (m.getName().equals(methodName)) {
                        method = m;
                    }
                }

                if (method != null) {
                    Class<? extends AbstractMessage> argumentType = (Class<? extends AbstractMessage>) method.getParameterTypes()[0];

                    client.postCallback(new ServiceMethodNotification(argumentType, packetMsg));
                }
            } catch (ClassNotFoundException e) {
                // The RPC service implementation was not implemented.
                // Either the .proto is missing, or the service was not converted to an interface yet.
                logger.error("Service Method: " + serviceName + ", was not found");
            }
        }
    }
}

File: src/main/java/in/dragonbra/javasteam/steam/handlers/steamunifiedmessages/SteamUnifiedMessages.kt
package `in`.dragonbra.javasteam.steam.handlers.steamunifiedmessages

import com.google.protobuf.AbstractMessage
import com.google.protobuf.GeneratedMessage
import `in`.dragonbra.javasteam.base.ClientMsgProtobuf
import `in`.dragonbra.javasteam.base.IPacketMsg
import `in`.dragonbra.javasteam.base.PacketClientMsgProtobuf
import `in`.dragonbra.javasteam.enums.EMsg
import `in`.dragonbra.javasteam.steam.handlers.ClientMsgHandler
import `in`.dragonbra.javasteam.steam.handlers.steamunifiedmessages.callback.ServiceMethodNotification
import `in`.dragonbra.javasteam.steam.handlers.steamunifiedmessages.callback.ServiceMethodResponse
import `in`.dragonbra.javasteam.types.AsyncJobSingle
import java.util.concurrent.ConcurrentHashMap

/**
 * @author Lossy
 * @since 2024-10-22
 *
 * This handler is used for interacting with Steamworks unified messaging.
 */
class SteamUnifiedMessages : ClientMsgHandler() {

    internal val handlers = ConcurrentHashMap<String, UnifiedService>()

    /**
     * Creates a service that can be used to send messages and receive notifications via Steamworks unified messaging.
     * @param TService The type of the service to create.
     * @return The instance to create requests from.
     */
    // Kotlin Compat
    @JvmSynthetic
    inline fun <reified TService : UnifiedService> createService(): TService = createService(TService::class.java)

    /**
     * Creates a service that can be used to send messages and receive notifications via Steamworks unified messaging.
     * @param TService The type of the service to create.
     * @param serviceClass The type of the service to create.
     * @return The instance to create requests.
     */
    // Java Compat
    @Suppress("UNCHECKED_CAST")
    fun <TService : UnifiedService> createService(serviceClass: Class<TService>): TService =
        serviceClass.getDeclaredConstructor(SteamUnifiedMessages::class.java)
            .newInstance(this@SteamUnifiedMessages)
            .let { service ->
                handlers.getOrPut(service.serviceName) { service } as TService
            }

    /**
     * Removes a service so it no longer can be used to send messages or receive notifications.
     * @param TService The type of the service to remove.
     */
    // Kotlin Compat
    @Suppress("unused")
    @JvmSynthetic
    inline fun <reified TService : UnifiedService> removeService() {
        removeService(TService::class.java)
    }

    /**
     * Removes a service so it no longer can be used to send messages or receive notifications.
     * @param TService The type of the service to remove.
     * @param serviceClass The type of the service to remove.
     */
    fun <TService : UnifiedService> removeService(serviceClass: Class<TService>) {
        val serviceName = serviceClass.getDeclaredConstructor(SteamUnifiedMessages::class.java)
            .newInstance(null)
            .serviceName
        handlers.remove(serviceName)
    }

    /**
     * Sends a message.
     * Results are returned in a [ServiceMethodResponse] with type [TResult].
     * The returned [AsyncJobSingle] can also be awaited to retrieve the callback result.
     * @param TRequest The type of protobuf object.
     * @param TResult The type of the result of the request.
     * @param name Name of the RPC endpoint. Takes the format ServiceName.RpcName.
     * @param message The message to send.
     * @return The JobID of the request. This can be used to find the appropriate [ServiceMethodResponse].
     */
    @Suppress("UNUSED_PARAMETER")
    fun <TRequest : GeneratedMessage.Builder<TRequest>, TResult : GeneratedMessage.Builder<TResult>> sendMessage(
        responseClass: Class<out TResult>, // Type Casting
        name: String,
        message: GeneratedMessage,
    ): AsyncJobSingle<ServiceMethodResponse<TResult>> {
        val eMsg = if (client.steamID == null) {
            EMsg.ServiceMethodCallFromClientNonAuthed
        } else {
            EMsg.ServiceMethodCallFromClient
        }

        val msg = ClientMsgProtobuf<TRequest>(message::class.java, eMsg).apply {
            sourceJobID = client.getNextJobID()
            header.proto.targetJobName = name
            body.mergeFrom(message)
        }

        client.send(msg)

        return AsyncJobSingle(client, msg.sourceJobID)
    }

    /**
     * Sends a notification.
     * @param TRequest The type of protobuf object.
     * @param name Name of the RPC endpoint. Takes the format ServiceName.RpcName.
     * @param message The message to send.
     */
    fun <TRequest : GeneratedMessage.Builder<TRequest>> sendNotification(
        name: String,
        message: GeneratedMessage,
    ) {
        // Notifications do not set source jobid, otherwise Steam server will actively reject this message
        // if the method being used is a "Notification"
        val eMsg = if (client.steamID == null) {
            EMsg.ServiceMethodCallFromClientNonAuthed
        } else {
            EMsg.ServiceMethodCallFromClient
        }
        val msg = ClientMsgProtobuf<TRequest>(message::class.java, eMsg).apply {
            header.proto.targetJobName = name
            body.mergeFrom(message)
        }

        client.send(msg)
    }

    /**
     * Handles a client message. This should not be called directly.
     * @param packetMsg The packet message that contains the data.
     */
    override fun handleMsg(packetMsg: IPacketMsg) {
        val packetMsgProto = packetMsg as? PacketClientMsgProtobuf ?: return

        if (packetMsg.msgType != EMsg.ServiceMethod && packetMsg.msgType != EMsg.ServiceMethodResponse) {
            return
        }

        val jobName = packetMsgProto.header.proto.targetJobName
        if (jobName.isEmpty()) return

        // format: Service.Method#Version
        val dot = jobName.indexOf('.')
        val hash = jobName.lastIndexOf('#')
        if (dot < 0 || hash < 0) return

        val serviceName = jobName.substring(0, dot)
        val handler = handlers[serviceName] ?: return
        val methodName = jobName.substring(dot + 1, hash)

        when (packetMsgProto.msgType) {
            EMsg.ServiceMethodResponse -> handler.handleResponseMsg(methodName, packetMsgProto)
            EMsg.ServiceMethod -> handler.handleNotificationMsg(methodName, packetMsgProto)
            else -> Unit // Ignore everything else.
        }
    }

    internal fun <TResponse : GeneratedMessage.Builder<TResponse>> handleResponseMsg(
        serviceClass: Class<out AbstractMessage>,
        packetMsg: PacketClientMsgProtobuf,
    ) {
        val callback = ServiceMethodResponse<TResponse>(serviceClass, packetMsg)
        client.postCallback(callback)
    }

    internal fun <TNotification : GeneratedMessage.Builder<TNotification>> handleNotificationMsg(
        serviceClass: Class<out AbstractMessage>,
        packetMsg: PacketClientMsgProtobuf,
    ) {
        val callback = ServiceMethodNotification<TNotification>(serviceClass, packetMsg)
        client.postCallback(callback)
    }
}
